bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java

```diff
@@ -53,38 +56,77 @@
     private static final String SELECTOR_FILTER_PROPERTY = "selectorFilter";
     private static final String INCLUDES_SUBCATEGORIES_PROPERTY = "includesSubCategories";
     private static final String ID_AND_URL_PATH_SEPARATOR = "|";
+    private static final String TEMPLATE_PREFIX = "{{";
+    private static final String TEMPLATE_SUFFIX = "}}";
 
     private String productUrlTemplate;
     private Pair<IdentifierLocation, ProductIdentifierType> productIdentifierConfig;
 
     private String categoryUrlTemplate;
-    private Pair<IdentifierLocation, CategoryIdentifierType> categoryIdentifierConfig;
+    private IdentifierLocation categoryIdentifierLocation;
+
+    private String queryParameterName;
 
     @Activate
     public void activate(UrlProviderConfiguration conf) {
         productUrlTemplate = conf.productUrlTemplate();
         productIdentifierConfig = Pair.of(conf.productIdentifierLocation(), conf.productIdentifierType());
 
         categoryUrlTemplate = conf.categoryUrlTemplate();
-        categoryIdentifierConfig = Pair.of(conf.categoryIdentifierLocation(), conf.categoryIdentifierType());
+        categoryIdentifierLocation = conf.categoryIdentifierLocation();
+
+        queryParameterName = conf.identifierQueryParamater();
     }
 
     @Override
     public String toProductUrl(SlingHttpServletRequest request, Page page, Map<String, String> params) {
         return toUrl(request, page, params, productUrlTemplate);
     }
 
+    @Override
+    public String toProductUrl(SlingHttpServletRequest request, Page page, String productIdentifier, MagentoGraphqlClient graphqlClient) {
```

Comment:
Is adding the graphql client here really necessary? From a first impression it feels like a graphql client does not really belong to url generation. 

Would it maybe make sense to adapt the MagentoGraphqlClient from the request (as a cached model) instead?

Reply:
> Is adding the graphql client here really necessary? From a first impression it feels like a graphql client does not really belong to url generation.

I think yes & no. URLProvider needs product (or category) specific information to generate to URL. In the past we had the responsibiliy to provided these by the components (and I keeped this pattern). This is something I want to avoid because:
* components must have knowledge about the configured URL pattern and variables used 
* it forces components to query extra attributes and some to have a query at all

Reply:
Related to "adapt the MagentoGraphqlClient from the request" ... as discuss in context of CIF-2137 we can adapt from request, once this is possible.

Reply:
@Buuhuu solved, `graphqlClient` has been removed and is adapted internal from the request.